---
title: A generalized linear mixed model for miRNA-seq data
author: '[Andrea M. Baran and Matthew N. McCall](http://mnmccall.com)'
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteEngine{knitr::knitr}
  %\VignetteIndexEntry{A generalized linear mixed model for miRNA-seq data}
  %\usepackage[UTF-8]{inputenc}
---


```{r global_options, include=FALSE}
library(knitr)
knitr::opts_chunk$set(fig.width=12, fig.height=8,
                      warning=FALSE, message=FALSE)
```

# Introduction

In this vignette we demonstrate how to analyze microRNA-seq data at the isomiR-level using the miRglmm package. This package begins with a SummarizedExperiment object containing aligned isomiR-level read counts. The alignment can be performed by either miRge or srabench. We provide instructions for both alignment algorithms, starting with the fastq files and concluding with the SummarizedExperiment object to be used as input to the miRglmm function for modeling. Each step in the analysis is briefly described and carried out by one or more functions implemented in this package. Each of these functions has its own help file describing it in detail; here we focus on the analysis as a whole. 

# Experimental Design
Our example data set consists of 6 monocyte samples. Fastq files for these samples can be accessed through the sequence read archive (SRA), via this link:


# miRge processing and alignment
Arun: add miRge instructions here

# sRNAbench processing and alignment
After installing _sRNAbench_ ([instructions here](https://github.com/bioinfoUGR/sRNAtoolbox)), you can directly download and align SRA files from your terminal. To process the 6 monocytes samples from [SRP110505](https://www.ncbi.nlm.nih.gov/sra?term=SRP110505) included as example data in this package:
```
sRNAbench input=SRR5755870 output=srnabench/SRR5755870 adapter=TGGAATTCTCGGGTGCCAAGGG microRNA=hsa isoMiR=true
sRNAbench input=SRR5755996 output=srnabench/SRR5755996 adapter=TGGAATTCTCGGGTGCCAAGGG microRNA=hsa isoMiR=true
sRNAbench input=SRR5755961 output=srnabench/SRR5755961 adapter=TGGAATTCTCGGGTGCCAAGGG microRNA=hsa isoMiR=true
sRNAbench input=SRR5755982 output=srnabench/SRR5755982 adapter=TGGAATTCTCGGGTGCCAAGGG microRNA=hsa isoMiR=true
sRNAbench input=SRR5755821 output=srnabench/SRR5755821 adapter=TGGAATTCTCGGGTGCCAAGGG microRNA=hsa isoMiR=true
sRNAbench input=SRR5756084 output=srnabench/SRR5756084 adapter=TGGAATTCTCGGGTGCCAAGGG microRNA=hsa isoMiR=true
```
We then run _sRNAde_ to generate a matrix of counts:
```
sRNAde input=srnabench/ output=sRNAde_result grpString=SRR5755870:SRR5755996:SRR5755961:SRR5755982:SRR5755821:SRR5756084 readIso=true colReadIso=5 minExprReadIso=1
```
The resulting isomiR count matrix file, _microRNAannotation_minExpr1_RC.mat_,  can be found in the _sRNAde_result_ directory.
# Statistical modeling with miRglmm

```{r load_miRglmm}
library("miRglmm")
library("SummarizedExperiment")
```

```{r load_example_data}
data("se_example")
```

The example data and corresponding annotation are stored in a SummarizedExperiment object:

```{r data_explore}
show(se_example)
str(assay(se_example))

```

The column names of the assay in the SummarizedExperiment object must correspond to unique sample IDs:

```{r data_explore2}

head(colnames(assay(se_example)))

```

The colData contain information about each sample:
```{r col_data_explore}
colData(se_example)

```

The rowData contain information about each isomiR, including its sequence ("uniqueSequence") and the miRNA ("miRNA") it aligned to:
```{r row_data_explore}
rowData(se_example)
```

## Fitting the GLMM model
A SummarizedExperiment object (for example, se_example) is input into miRglmm, along with a vector specifying primary fixed effect of interest for Differential Expression Analysis (DEA), col_group. This can be 2 or more level categorical variable. For miRglmm to appropriately model the random effects for sample and sequence, the columns names of the assay object in the input SummarizedExperiment object (for example, assay(se_example)) must be the unique sample IDs. There must also be elements of rowData (for example, rowData(se_example)) that are called "uniqueSequence" and "miRNA", which indicate the unique sequence/isomiR identifier, and the unique miRNA identifiers, respectively.

Additional input arguments include: min_med_lcpm (default -1), ncores (default 1), and adjust_var (default NA). min_med_lcpm specifies a threshold on median(log(CPM)) expression for retaining sequences to include in the analysis. Larger values retain fewer (more highly expressed) sequences/isomiRs in the analysis. ncores is used to specify the number of cores available to run the algorithm, and the analysis of individual miRNA will be run in parallel across cores if ncores>1. adjust_var can be used to specify an adjustment variable (continuous or categorical) that will be used as an additional fixed effect in the miRglmm model. 

```{r miRglmm_fit}
fit = miRglmm(se = se_example, col_group = rep(c("A","B"), c(3,3)))
```

## Output of the GLMM model
The output of miRglmm is a list of glmerMod model fit objects called "miRglmm" with length equal to the number of miRNA modelled. The elements of the list are named by miRNA (based on unique entries in rowData(se)$miRNA). This allows flexible extraction of any model fit parameter of interest for downstream analysis. 

An additional output of miRglmm is another list of glmerMod model fits called "miRglmm_reduced". This is equivalent in size and structure to the "miRglmm" list output, but it is model fit objects after removal of the col_group|sequence random effect, and can be used to perform a likelihood ratio test for the significance of the random slope parameter.

## Downstream analyses 
The fixed effects can be obtained using the following commands:

```{r get_beta}
all_coeff=sapply(fit[["miRglmm"]], "fixef")
```

The effect of primary interest (col_group) can be obtained by looking at that specific row in the fit object:

```{r get_beta2}
library(stringr)
idx=which(str_detect(rownames(all_coeff), "col_group"))
coeff_full=data.frame("full"=all_coeff[idx,])
rownames(coeff_full)=colnames(all_coeff)
```

Plots of the primary fixed effect of interest, with overlaid isomiR-level random effects, can be plotted for a specified miRNA. Here we use hsa-miR-100-5p as an example:
```{r beta_plot}
miRNA_plot="hsa-miR-100-5p"

f1=fit[["miRglmm"]][[miRNA_plot]]

#pull out individual sequence estimates
idx=which(str_detect(colnames(ranef(f1)$sequence), "col_group"))
idx_fixed=which(str_detect(names(fixef(f1)), "col_group"))
groupA=fixef(f1)[1]+ranef(f1)$sequence$`(Intercept)`-fixef(f1)[1]
groupB=fixef(f1)[1]+ranef(f1)$sequence$`(Intercept)`+ranef(f1)$sequence[,idx]+fixef(f1)[idx_fixed]-fixef(f1)[1]
groupA=data.frame("estimate"=groupA, "sequence"=rownames(ranef(f1)$sequence))
groupB=data.frame("estimate"=groupB, "sequence"=rownames(ranef(f1)$sequence))
groupA$Pool="Bladder"
groupB$Pool="Testes"
comb_df=rbind(groupA, groupB)

#pull out fixed effect estimates
groupA=fixef(f1)[1]-fixef(f1)[1]
groupB=fixef(f1)[1]+fixef(f1)[idx_fixed]-fixef(f1)[1]
groupA=data.frame("estimate"=groupA)
groupB=data.frame("estimate"=groupB)
groupA$Pool="Bladder"
groupB$Pool="Testes"
comb_df_overall=rbind(groupA, groupB)

ggplot(comb_df, aes(x=Pool, y=estimate, group=sequence, color="miRglmm isomiR estimates (random effects)"))+geom_line(alpha=0.5)+geom_point(alpha=0.5)+
  geom_line(data=comb_df_overall, aes(x=Pool, y=estimate, group=1, color="miRglmm estimates (fixed effects)"), size=2)+
  xlab("tissue")+
  scale_y_continuous(breaks=c(-4, -2, 0, 2), labels=round(exp(c(-4,-2,0,2)),3))+
  ylab(paste('Expression relative to', '\n', 'average bladder expression'))+scale_x_discrete(expand=c(0.05,0))+xlab('')+
  scale_color_manual(name="", values=c("miRglmm estimates (fixed effects)"="red", "miRglmm isomiR estimates (random effects)"="black"))+
  theme(legend.position="bottom", legend.direction="horizontal")

```



The Standard Error (SE), and 95% Wald-based Confidence Intervals, for the primary fixed effect of interest can be found via:

```{r get_SE}
all_SE=sapply(fit[["miRglmm"]], "vcov")
idx1=which(str_detect(rownames(all_SE[[1]]), "col_group")==TRUE)
idx2=which(str_detect(colnames(all_SE[[1]]), "col_group")==TRUE)
SE_vec=data.frame('SE_full'=sapply(all_SE, function(x) sqrt(x[idx1,idx2])))
rownames(SE_vec)=names(all_SE)

z_alpha=qnorm(1-(1-0.95)/2)
LL_mat=coeff_full-z_alpha*SE_vec
UL_mat=coeff_full+z_alpha*SE_vec
```

Wald p-values can be found with the following:

```{r get_p}
all_pvals=sapply(fit[["miRglmm"]], function(f) summary(f)$coefficients[, "Pr(>|z|)"])
idx=which(str_detect(rownames(all_pvals), "col_group"))
pval_full=data.frame("full"=all_pvals[idx,])
rownames(pval_full)=colnames(all_pvals)
```

A likelihood-ratio test for the col_group|sequence effect can be performed via:
```{r LRT}
uniq_miRNA=intersect(names(fit[["miRglmm"]]), names(fit[["miRglmm_reduced"]]))
LRTp=data.frame("LRTp"=sapply(uniq_miRNA, function(row) anova(fit[["miRglmm"]][[row]], fit[["miRglmm_reduced"]][[row]])$`Pr(>Chisq)`[2]))
rownames(LRTp)=uniq_miRNA
```

The distribution of the LRT p-values can be plotted via:
```{r LRT_plot}
library(ggplot2)
ggplot(LRTp, aes(x=LRTp))+geom_histogram(color="black", fill="gray", bins=50)+xlab('Likelihood Ratio Test p-value')+ylab('number of miRNA')
```
The following code can be used to obtain the variance components from the (full) miRglmm model:
```{r var_comp}
uniq_miRNA=names(fit[["miRglmm"]])
  var_comp=data.frame('random_int_sample_var'=sapply(uniq_miRNA, function(row) VarCorr(fit[["miRglmm"]][[row]])$sample_labels[1,1]))
  var_comp2=data.frame('random_int_seq_var'=sapply(uniq_miRNA, function(row) VarCorr(fit[["miRglmm"]][[row]])$sequence[1,1]))
  var_comp=transform(merge(var_comp, var_comp2, by='row.names', all=T), row.names=Row.names, Row.names=NULL)
  var_comp2=data.frame('random_slope_seq_var'=sapply(uniq_miRNA, function(row) VarCorr(fit[["miRglmm"]][[row]])$sequence[2,2]))
  var_comp=transform(merge(var_comp, var_comp2, by='row.names', all=T), row.names=Row.names, Row.names=NULL)
```


# Session Info
```{r session_info}
sessionInfo()
```
